pkg/identity/gitlabcom/principal.go

```diff
@@ -0,0 +1,205 @@
+	repository string
+
+	// ID to the source repo
+	repositoryID string
+
+	// Owner of the source repo (mutable)
+	repositoryOwner string
+
+	// ID of the source repo
+	repositoryOwnerID string
+
+	// job ID
+	jobID string
+
+	// Git ref being built
+	ref string
+
+	// Commit SHA being built
+	sha string
+
+	// ID of the runner
+	runnerID int64
+
+	// The type of runner used by the job. May be one of gitlab-hosted or self-hosted.
+	runnerEnvironment string
+}
+
+func JobPrincipalFromIDToken(_ context.Context, token *oidc.IDToken) (identity.Principal, error) {
+	var claims struct {
+		ProjectPath       string `json:"project_path"`
+		ProjectID         string `json:"project_id"`
+		PipelineSource    string `json:"pipeline_source"`
+		PipelineID        string `json:"pipeline_id"`
+		NamespacePath     string `json:"namespace_path"`
+		NamespaceID       string `json:"namespace_id"`
+		JobID             string `json:"job_id"`
+		Ref               string `json:"ref"`
+		RefType           string `json:"ref_type"`
+		Sha               string `json:"sha"`
+		RunnerEnvironment string `json:"runner_environment"`
+		RunnerID          int64  `json:"runner_id"`
+	}
+
+	if err := token.Claims(&claims); err != nil {
+		return nil, err
+	}
+
+	if claims.ProjectPath == "" {
+		return nil, errors.New("missing project_path claim in ID token")
+	}
+
+	if claims.PipelineSource == "" {
+		return nil, errors.New("missing pipeline_source claim in ID token")
+	}
+
+	if claims.PipelineID == "" {
+		return nil, errors.New("missing pipeline_id claim in ID token")
+	}
+
+	if claims.JobID == "" {
+		return nil, errors.New("missing job_id claim in ID token")
+	}
+
+	if claims.Ref == "" {
+		return nil, errors.New("missing ref claim in ID token")
+	}
+
+	if claims.RefType == "" {
+		return nil, errors.New("missing ref_type claim in ID token")
+	}
+
+	if claims.NamespacePath == "" {
+		return nil, errors.New("missing namespace_path claim in ID token")
+	}
+
+	if claims.NamespaceID == "" {
+		return nil, errors.New("missing namespace_id claim in ID token")
+	}
+
+	if claims.ProjectID == "" {
+		return nil, errors.New("missing project_id claim in ID token")
+	}
+
+	if claims.Sha == "" {
+		return nil, errors.New("missing sha claim in ID token")
+	}
+
+	if claims.RunnerEnvironment == "" {
+		return nil, errors.New("missing runner_environment claim in ID token")
+	}
+
+	if claims.RunnerID == 0 {
+		return nil, errors.New("missing runner_id claim in ID token")
+	}
+
+	var ref string
+	switch claims.RefType {
+	case "branch":
+		ref = "refs/heads/" + claims.Ref
+	case "tag":
+		ref = "refs/tags/" + claims.Ref
+	default:
+		return nil, fmt.Errorf("unexpected ref_type: %s", claims.RefType)
+	}
+
+	return &jobPrincipal{
+		subject:           token.Subject,
+		issuer:            token.Issuer,
+		url:               `https://gitlab.com/`,
+		eventName:         claims.PipelineSource,
+		pipelineID:        claims.PipelineID,
+		repository:        claims.ProjectPath,
```

Comment:
In GitLab, we distinguish between projects and repositories. https://gitlab.com/gitlab-org/gitlab is a URL to a project, https://gitlab.com/gitlab-org/gitlab.git is a URL to a repository. A project contains a repository, along with other things like issues, merge requests, pipelines, etc.

TL;DR: `project` is the correct thing to call this, but `repository` is ok if we want to keep consistency with the other principals.